sql/schemachanger/scexec/scmutationexec: bug fix in addNewIndexMutation

Xiang-Gu opened the pull request:
The first two commits are some small changes and the third commit is the main commit.

Commit 1: Change the signature of an existing function
Commit 2: Add `ConstraintID` as an attribute to `TemporaryIndex` element 
in `screl`.
Commit 3 (the main commit): See below for a detailed description.

Previously, we call `addNewIndexMutation` to enqueue an index mutation
    to the mutation slice. However, that function is buggy in that it
    ignored the `constraintID` field of the input `opIndex` and instead
    called `tbl.NextConstraintID()` and assign it to the index descriptor.
    This is problematic because in the declarative schema changer world, we
    strive to uphold a rule that we shall manage column/index/constraint
    ID allocation in the builder, where we have access to the builder state
    information, so we know what the right next column/index/constraint ID
    in a table is.

To accompany that rule, in the mutation operation (such as this function
    `addNewIndexMutation`), we ought to check for column/index/constraint
    ID of the input element and update the table's
    `nextColumn/Index/ConstraintID` accordingly.

Such a change alo discovered and fixed two other existing issues:
       1. In `CREATE INDEX` and `ADD COLUMN`, we did not allocate
       constraintID to the new primary index or new unique secondary
       index that are created in the builder.
       2. We `--rewrite`'ed a bunch of test output pertaining to
       constraintID of indexes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85783)
<!-- Reviewable:end -->

Xiang-Gu committed:
sql/schemachanger: Change the signature of `b.nextTableConstraintID`

Changed the input argument of `b.nextTableConstraintID` from
`*scpb.Table` to a `catid.DescID`, as this turns out to be easier to
use in various places.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1, 30 of 30 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85783#-:-N9HpK-W6LMjx83IxLPz:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Xiang-Gu)



<!-- Sent from Reviewable.io -->

Xiang-Gu committed:
sql/schemachanger/screl: added attribute `ConstraintID` to `TempIndex`

postamar commented:
Great find!

Xiang-Gu committed:
sql/schemachanger/scexec/scmutationexec: bug fix in addNewIndexMutation

Previously, we call addNewIndexMutation to enqueue an index mutation
to the mutation slice. However, that function is buggy in that it
ignored the constraintID field of the input opIndex and instead
called tbl.NextConstraintID() and assign it to the index descriptor.
This is problematic because in the declarative schema changer world, we
strive to uphold a rule that we shall manage column/index/constraint
ID allocation in the builder, where we have access to the builder state
information, so we know what the right next column/index/constraint ID
in a table is.

To accompany that rule, in the mutation operation (such as this function
addNewIndexMutation), we ought to check for column/index/constraint
ID of the input element and update the table's
nextColumn/Index/ConstraintID accordingly.

Such a change alo discovered and fixed two other existing issues:
1. In CREATE INDEX and ADD COLUMN, we did not allocate
constraintID to the new primary index or new unique secondary
index that are created in the builder.
2. We --rewrite'ed a bunch of test output pertaining to
constraintID of indexes.

Release note: None

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 20 of 217 files at r3, 5 of 30 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85783#-:-N9IXIxj3zPgMZSHn5h5:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

Xiang-Gu commented:
TFTR!

bors r+

Xiang-Gu commented:
bors merge conflict does not seem to come from this PR, retrying

bors r+

Xiang-Gu commented:
retrying bors after a merge conflict:

bors r+
